### PR TITLE
Remove Licence from bill 2PT bill run during review

### PR DESCRIPTION
**Pull request by Jozzey**
https://eaflood.atlassian.net/browse/WATER-4216

On the licence review page add a button for “remove from bill run”. When clicked you should be presented with a confirmation page with a confirm button, when this is clicked the licence should be removed from the current bill run.

Once the licence is removed the user should be returned to the bill run review page with a banner confirming the licence has been removed. When removed the licence will be flagged for two-part tariff supplementary billing.

### CODE REVIEW

File: app/services/bill-runs/two-part-tariff/review-bill-run.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/review-bill-run.service.js:18]: NIT

```suggestion
 * @param {Object} yar - The Hapi `request.yar` session manager passed on by the controller
 *
```

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:18]: Our convention for the documentation is to have a single line, then a space, a more detailed description (optional) and then the params. And if the function returns something that needs to be included.

```suggestion
 * Orchestrates removing a licence from a bill run whilst it is at the review stage
 *
 * It does this by deleting all of the persisted data relating to the licence from the review tables. The licence will
 * then be flagged for 2PT supplementary billing. If after removing a licence the bill run is empty, the bill run status
 * will be set to `empty` and `true` returned so that the user is redirected back to the Bill runs page rather
 * than Review bill run.
 *
 * @param {string} billRunId - The UUID of the bill run that the licence is in
 * @param {string} licenceId - UUID of the licence to remove from the bill run
 * @param {Object} yar - The Hapi `request.yar` session manager passed on by the controller
 *
 * @returns {Promise<boolean>} true if all the licences have been removed from the bill run else false
```

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:26]: I'd suggest moving these off into a separate function, for example, `_removeReviewData()`.

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:40]: If we use the Objection model we can replace this with 1 line

```suggestion
  const count = await ReviewLicenceModel.query().where('billRunId', billRunId).resultSize()
```

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:49]: If we use the Objection model we can replace this with 1 line

```suggestion
  if (count === 0) {
    await BillRunModel.query().findById(billRunId).patch({ status: 'empty' })

    return true
  }
```

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:54]: We shouldn't be doing this in this method as it is a clear violation of 'Single responsibility principle' from [SOLID](https://en.wikipedia.org/wiki/SOLID).

We can use the result this function returns in the `go()` to determine if the flash banner needs setting. That also means we do not need to pass yar or the licence ref into this function.

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Cruikshanks) [inline app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:126]: If we use the Objection model we can replace this with 1 line. I would also suggest a change to the function name

```suggestion
async function _flagForSupplementaryBilling (licenceId) {
  const licence = await LicenceModel.query().findById(licenceId).patch({ includeInSrocTptBilling: true })

  return licence.licenceRef
}
```

File: app/views/bill-runs/remove-licence.njk
Comment (Cruikshanks) [inline app/views/bill-runs/remove-licence.njk:56]: The `preventDoubleClick` option needs to be enabled on the button

```suggestion
        {{ govukButton({
          text: "Remove licence",
          preventDoubleClick: true
        }) }}
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
I've not got into the tests because I figure it would better to hold off until some of these suggested changes are reviewed and we get an answer back from our design team.

**Comment by Cruikshanks** [comment]
I've put this on-hold because I've seen that the intended 'confirm remove licence from bill run' is quite different from the one that already exists and is being used.

I'd prefer that we keep things consistent but they'll get the final say. We just need to double-check what they want to do. Like most things, they've probably forgotten or are not aware the page already exists!

File: app/presenters/bill-runs/two-part-tariff/remove-bill-run-licence.presenter.js
Comment (Jozzey) [reply app/presenters/bill-runs/two-part-tariff/remove-bill-run-licence.presenter.js:20]: It can be an array, but the array will only contain data for a single licence. There will only be multiple elements if a licence has more than one billing account. So I would say calling it licence is correct.

File: app/services/bill-runs/two-part-tariff/review-bill-run.service.js
Comment (Jozzey) [reply app/services/bill-runs/two-part-tariff/review-bill-run.service.js:18]: Yes I didn't like that either :grin:. I copied it from the original one you did in `app/services/bill-runs/two-part-tariff/review-licence.service.js` so figured it was your convention for `yar`

Comment (Jozzey) [reply app/services/bill-runs/two-part-tariff/review-bill-run.service.js]: I've also sorted that out in `app/services/bill-runs/two-part-tariff/review-licence.service.js`

File: app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js
Comment (Jozzey) [reply app/services/bill-runs/two-part-tariff/submit-remove-bill-run-licence.service.js:125]: I'll make this change to use objection. Just to note though that for it to return the `licenceRef` I'll need to add `.returning('licenceRef')` to the end of it otherwise it just returns the number of records updated. This also means that it is no longer a 1 liner as it goes over the line length.

File: app/presenters/bill-runs/two-part-tariff/remove-bill-run-licence.presenter.js
Comment (Jozzey) [reply app/presenters/bill-runs/two-part-tariff/remove-bill-run-licence.presenter.js]: It is not an issue now that we are not displaying the billing account number

**Comment by Jozzey** [comment]
Taken off hold now we have the new page layout confirmed and built

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**
